AsyncDisplayKit/ASNetworkImageNode.mm

```diff
@@ -350,41 +348,56 @@ - (CGSize)calculateSizeThatFits:(CGSize)constrainedSize
   return [super calculateSizeThatFits:constrainedSize];
 }
 
-#pragma mark - Private methods -- only call with lock.
+#pragma mark - Private methods, safe to call without lock
 
-- (void)_updateProgressImageBlockOnDownloaderIfNeeded
+- (void)handleProgressImage:(UIImage *)progressImage progress:(CGFloat)progress downloadIdentifier:(nullable id)downloadIdentifier
 {
   ASDN::MutexLocker l(__instanceLock__);
-  
-  BOOL shouldRenderProgressImages = _shouldRenderProgressImages;
-  ASInterfaceState interfaceState = self.interfaceState;
+  // Getting a result back for a different download identifier, download must not have been successfully canceled
+  if (ASObjectIsEqual(_downloadIdentifier, downloadIdentifier) == NO && downloadIdentifier != nil) {
+    return;
+  }
+  self.image = progressImage;
+  dispatch_async(dispatch_get_main_queue(), ^{
+    // See comment in -displayDidFinish for why this must be dispatched to main
+    self.currentImageQuality = progress;
+  });
+}
 
-  if (!_downloaderFlags.downloaderImplementsSetProgress || _downloadIdentifier == nil) {
+#pragma mark - Private methods, call with lock held
+
+- (void)_updateProgressImageBlockOnDownloaderIfNeeded
+{
+  // If the downloader doesn't do progress, we are done.
+  if (_downloaderFlags.downloaderImplementsSetProgress == NO) {
```

Comment:
No. Even if `_downloadIdentifier == nil` we still want to continue. We may need to disconnect from the prior download identifier.